### PR TITLE
[llvm][AArch64] Do not inline a function with different signing scheme.

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
If the signing scheme is different that maybe the functions assumes different behaviours and dangerous to inline them without analysing them. This should be a rare case.

### CODE REVIEW

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 1]
```diff
@@ -5155,7 +5155,39 @@ struct StrictFPUpgradeVisitor : public InstVisitor<StrictFPUpgradeVisitor> {
 };
 } // namespace
 
-void llvm::UpgradeFunctionAttributes(Function &F) {
+static void
+CopyModuleAttributeToFunction(Function &F, StringRef FnAttrName,
+                              StringRef ModAttrName,
+                              std::pair<StringRef, StringRef> Values) {
```
Comment (nickdesaulniers) [inline llvm/lib/IR/AutoUpgrade.cpp:5161]: Can you add a comment about what the pair of StringRef are supposed to refer to? The parameter identifier of Values doesn't betray what the caller should be passing here.

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 2]
```diff
@@ -5155,7 +5155,39 @@ struct StrictFPUpgradeVisitor : public InstVisitor<StrictFPUpgradeVisitor> {
...
+  Module *M = F.getParent();
+  if (!M)
+    return;
```
Comment (nickdesaulniers) [inline llvm/lib/IR/AutoUpgrade.cpp:5164]: Consider making this an assertion unless you observed cases where this was false?  The call chain checks this already in `llvm::UpgradeFunctionAttributes`, so perhaps an assertion here in the callee is more appropriate, in case this function grows more callers in the future.

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 3]
```diff
@@ -5155,7 +5155,39 @@ struct StrictFPUpgradeVisitor : public InstVisitor<StrictFPUpgradeVisitor> {
...
+  if (F.hasFnAttribute(FnAttrName))
+    return;
```
Comment (nickdesaulniers) [inline llvm/lib/IR/AutoUpgrade.cpp:5166]: Could check this first before checking for the parent module.

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 4]
```diff
@@ -5155,7 +5155,39 @@ struct StrictFPUpgradeVisitor : public InstVisitor<StrictFPUpgradeVisitor> {
...
+  CopyModuleAttributeToFunction(
+      F, AttrName, AttrName,
+      std::make_pair<StringRef, StringRef>("true", "false"));
+}
+
+static void
+CopyModuleAttributeToFunction(Function &F, StringRef AttrName,
+                              std::pair<StringRef, StringRef> Values) {
+  CopyModuleAttributeToFunction(F, AttrName, AttrName, Values);
+}
```
Comment (nickdesaulniers) [inline llvm/lib/IR/AutoUpgrade.cpp:5187]: Please add function level comments for these two describing when one should be used vs the other.

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 5]
```diff
@@ -5167,6 +5199,41 @@ void llvm::UpgradeFunctionAttributes(Function &F) {
...
+    if (!F.hasFnAttribute("sign-return-address")) {
+      StringRef SignType = "none";
+      if (const auto *Sign = mdconst::extract_or_null<ConstantInt>(
+              M->getModuleFlag("sign-return-address")))
+        if (Sign->getZExtValue())
+          SignType = "non-leaf";
+
+      if (const auto *SignAll = mdconst::extract_or_null<ConstantInt>(
+              M->getModuleFlag("sign-return-address-all")))
+        if (SignAll->getZExtValue())
```
Comment (nickdesaulniers) [inline llvm/lib/IR/AutoUpgrade.cpp:5225]: This pattern appears a lot; consider using a small static helper function.

File: llvm/lib/Linker/IRMover.cpp [hunk 6]
```diff
@@ -1606,6 +1606,10 @@ Error IRLinker::run() {
   // Loop over all of the linked values to compute type mappings.
   computeTypeMapping();
 
+  // Update function attributes before copy them to destation module.
```
Comment (nickdesaulniers) [inline llvm/lib/Linker/IRMover.cpp:1609]: s/copy/copying/

File: llvm/lib/Transforms/Utils/InlineFunction.cpp [hunk 7]
```diff
@@ -2110,6 +2110,21 @@ llvm::InlineResult llvm::InlineFunction(CallBase &CB, InlineFunctionInfo &IFI,
...
+  if (CalledFunc->getFnAttribute("sign-return-address-key") !=
+      Caller->getFnAttribute("sign-return-address-key")) {
+    return InlineResult::failure("incompatible signing keys attributes");
+  }
+  if (CalledFunc->getFnAttribute("branch-protection-pauth-lr") !=
+      Caller->getFnAttribute("branch-protection-pauth-lr")) {
+    return InlineResult::failure(
+        "incompatible sign return address modifier attributes");
+  }
+
```
Comment (nickdesaulniers) [inline llvm/lib/Transforms/Utils/InlineFunction.cpp:2127]: I don't think we should be adding more explicit checks like this.

I think `CompatRule` in llvm/include/llvm/IR/Attributes.td should be used instead. The `StrictFP` check above could be converted to a `CompatRule` instead.

Can you please give that a shot?

These rules are checked earlier by `AttributeFuncs::areInlineCompatible` in InlineCost.cpp.

File: llvm/test/LTO/AArch64/link-sign-return-address.ll [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+; CHECK-DUMP:     autiasp
+; CHECK-DUMP:     ret
+; CHECK-DUMP: <main>:
+; CHECK-DUMP-NOT:  paciasp
+; CHECK-DUMP:      str     x30,
+; CHECK-DUMP:      bl      0x14 <main+0x4>
+
+; `main` doesn't support PAC sign-return-address while `foo` does, so in the binary
+; we should not see anything.
+; CHECK-PROP-NOT:   Proper ties: aarch64 feature: PAC
```
Comment (nickdesaulniers) [inline llvm/test/LTO/AArch64/link-sign-return-address.ll:43]: is there an extra whitespace in `Proper` and `ties`?

File: llvm/test/Transforms/Inline/inline-sign-return-address.ll [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  ret i32 46
+}
+
+define dso_local i32 @bar_all() #0 {
+; CHECK-LABEL: bar_all
+; CHECK-NOT:   call i32 @foo_all()
+; CHECK:       call i32 @foo_nonleaf()
+; CHECK:       call i32 @foo_none()
+; CHECK:       call i32 @foo_lr()
+; CHECK:       call i32 @foo_bkey()
```
Comment (nickdesaulniers) [inline llvm/test/Transforms/Inline/inline-sign-return-address.ll:32]: Prefer CHECK-NEXT when applicable.

File: llvm/test/Transforms/Inline/inline-sign-return-address.ll [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  %1 = call i32 @foo_all()
+  %2 = call i32 @foo_nonleaf()
+  %3 = call i32 @foo_none()
+  %4 = call i32 @foo_lr()
+  %5 = call i32 @foo_bkey()
+  %6 = add nsw i32 %1, %2
+  %7 = add nsw i32 %6, %3
+  %8 = add nsw i32 %7, %4
+  %9 = add nsw i32 %8, %5
+  ret i32 %9
```
Comment (nickdesaulniers) [inline llvm/test/Transforms/Inline/inline-sign-return-address.ll:42]: These `add` are adding noise to the test.  Just make the called functions `void` functions, and don't bother with these adds.

**Comment by nickdesaulniers** [comment]
Also, consider submitting your 3 commits as 3 distinct PRs. The first is ready to land.  I think the auto upgrade could be detached from the inlining decisions.

Also, I'm kind of surprised you didn't have to change anything in clang codegen of LLVM IR.  Shouldn't clang now be generating fn attrs rather than module attrs?  Was there a different PR for that?

File: llvm/test/Transforms/Inline/inline-sign-return-address.ll (hunk 10 above)
Comment (DanielKristofKiss) [reply llvm/test/Transforms/Inline/inline-sign-return-address.ll]: ack

File: llvm/lib/Transforms/Utils/InlineFunction.cpp (hunk 7 above)
Comment (DanielKristofKiss) [reply llvm/lib/Transforms/Utils/InlineFunction.cpp]: done.

**Comment by DanielKristofKiss** [comment]
> Also, consider submitting your 3 commits as 3 distinct PRs. The first is ready to land. I think the auto upgrade could be detached from the inlining decisions.

Sorry, messed up this one as other patches went up separately already: 
https://github.com/llvm/llvm-project/pull/80640 ( I'll update that with the comment here)
https://github.com/llvm/llvm-project/pull/80465 

> Also, I'm kind of surprised you didn't have to change anything in clang codegen of LLVM IR. Shouldn't clang now be generating fn attrs rather than module attrs? Was there a different PR for that?

Clang can't really do much here as we add more functions to the module during later passes ( c++ exception handler, sanitisers especially). So the right thing is to derive the attributes from the module flag. Autoupgrade now do this at the last moment before the functions merged into the common module.

File: llvm/lib/IR/Attributes.cpp [hunk 11]
```diff
@@ -2039,12 +2039,24 @@ static bool checkDenormMode(const Function &Caller, const Function &Callee) {
   return false;
 }
 
+static bool checkStrictFP(const Function &Caller, const Function &Callee) {
+  // Do not inline strictfp function into non-strictfp one. It would require
+  // conversion of all FP operations in host function to constrained intrinsics.
+  return !(Callee.getAttributes().hasFnAttr(Attribute::StrictFP) &&
+           !Caller.getAttributes().hasFnAttr(Attribute::StrictFP));
+}
+
```
Comment (nickdesaulniers) [inline llvm/lib/IR/Attributes.cpp:2048]: Make a separate PR for this? And thanks for cleaning this up!

File: llvm/include/llvm/IR/Attributes.td [hunk 12]
```diff
@@ -359,7 +366,10 @@ def : CompatRule<"isEqual<ShadowCallStackAttr>">;
 def : CompatRule<"isEqual<UseSampleProfileAttr>">;
 def : CompatRule<"isEqual<NoProfileAttr>">;
 def : CompatRule<"checkDenormMode">;
-
+def : CompatRule<"checkStrictFP">;
+def : CompatRuleAttr<"isEqual", "sign-return-address">;
+def : CompatRuleAttr<"isEqual", "sign-return-address-key">;
+def : CompatRuleAttr<"isEqual", "branch-protection-pauth-lr">;
```
Comment (nickdesaulniers) [inline llvm/include/llvm/IR/Attributes.td:372]: I guess I'm surprised that you can't do:
```
def : CompatRule<"isEqual<sign-return-address>">;
```
Is this because these attributes are string attributes?  If so, consider adding more to the comment blocks to describe //when// a developer should use `CompatRule` vs `CompatRuleAttr`.  Perhaps s/CompatRuleAttr/CompatRuleStrAttr/ since the existing vanillla CompatRule is already concerned with function attributes in order to better distinguish that this new class is for //string// attributes.

File: llvm/utils/TableGen/Attributes.cpp [hunk 13]
```diff
@@ -87,7 +87,12 @@ void Attributes::emitFnAttrCompatCheck(raw_ostream &OS, bool IsStringAttr) {
 
   for (auto *Rule : CompatRules) {
     StringRef FuncName = Rule->getValueAsString("CompatFunc");
-    OS << "  Ret &= " << FuncName << "(Caller, Callee);\n";
+    StringRef AttrName = Rule->getValueAsString("AttrName");
+    if (AttrName.empty())
+      OS << "  Ret &= " << FuncName << "(Caller, Callee);\n";
+    else
+      OS << "  Ret &= " << FuncName << "(Caller, Callee, \"" << AttrName
+         << "\");\n";
```
Comment (nickdesaulniers) [inline llvm/utils/TableGen/Attributes.cpp:95]:
```suggestion
    OS << "  Ret &= " << FuncName << "(Caller, Callee";
    StringRef AttrName = Rule->getValueAsString("AttrName");
    if (!AttrName.empty())
      OS << ", \"" << AttrName << "\"";
    OS <<  ");\n";
```

File: llvm/include/llvm/IR/Attributes.td (hunk 12 above)
Comment (DanielKristofKiss) [reply llvm/include/llvm/IR/Attributes.td:372]: > Is this because these attributes are string attributes?
Correct, and string can't be a template param ( `isEqual<"sign-return-address">"` won't work)

File: llvm/include/llvm/IR/Attributes.td [hunk 14]
```diff
@@ -339,20 +339,25 @@ def UseSampleProfile : StrBoolAttr<"use-sample-profile">;
 def DenormalFPMath : ComplexStrAttr<"denormal-fp-math", [FnAttr]>;
 def DenormalFPMathF32 : ComplexStrAttr<"denormal-fp-math-f32", [FnAttr]>;
 
+// Attiribute compatiblity rules are generated to check the attribute of the
```
Comment (nickdesaulniers) [inline llvm/include/llvm/IR/Attributes.td:342]: s/Attiribute/Attribute/

File: llvm/include/llvm/IR/Attributes.td [hunk 15]
```diff
@@ -339,20 +339,25 @@ def UseSampleProfile : StrBoolAttr<"use-sample-profile">;
...
+// caller and callee and decide whether inlining should be allowed. CompatRule
+// and child classes are used for the rule generation. CompatRule takes only a
+// compare function which could be templated with the attribute type.
+// CompatRuleStrAttr takes the compare function and the string for checking the
```
Comment (nickdesaulniers) [inline llvm/include/llvm/IR/Attributes.td:346]: s/string/string attribute/

File: llvm/include/llvm/IR/Attributes.td [hunk 16]
```diff
@@ -339,20 +339,25 @@ def UseSampleProfile : StrBoolAttr<"use-sample-profile">;
...
+// string attributes.
```
Comment (nickdesaulniers) [inline llvm/include/llvm/IR/Attributes.td:347]: s/the string attributes/compatibility for inline substitution/.

**Review by nickdesaulniers — APPROVED** [review]
LGTM; thanks!

Might want to wait to land this until after we've resolved the other patch; right now, these aren't function attributes that are generated anywhere (until the other PR is resolved)

**Comment by DanielKristofKiss** [comment]
> LGTM; thanks!
> 
> Might want to wait to land this until after we've resolved the other patch; right now, these aren't function attributes that are generated anywhere (until the other PR is resolved)
(#80640) is landed so landing this too.
